File: 2327. Number of People Aware of a Secret/Solution.cpp
#include <deque>
#include <vector>
using namespace std;

class Solution {
public:
    // 动态规划
    int peopleAwareOfSecret_MK1(int n, int delay, int forget)
    {
        constexpr int MOD = 1e9 + 7;
        vector<long long> dp(n);
        dp[0] = 1;
        int slide_sum = 0;
        for (int i = 1; i < n; ++i) {
            if (i >= delay)
                slide_sum = (slide_sum + dp[i - delay]) % MOD;
            if (i >= forget)
                slide_sum = (slide_sum - dp[i - forget] + MOD) % MOD;
            dp[i] = slide_sum;
        }

        int ans = 0;
        for (int i = n - forget; i < n; ++i)
            ans = (ans + dp[i]) % MOD;
        return ans;
    }

    // 前缀和
    int peopleAwareOfSecret(int n, int delay, int forget)
    {
        constexpr int MOD = 1e9 + 7;
        vector<int> pre_sum(n + 1);
        pre_sum[0] = 0;
        pre_sum[1] = 1;
        for (int i = 2; i <= n; ++i) {
            int known = (pre_sum[max(0, i - delay)] - pre_sum[max(0, i - forget)]) % MOD;
            pre_sum[i] = (pre_sum[i - 1] + known) % MOD;
        }
        return (pre_sum[n] - pre_sum[max(0, n - forget)] + MOD) % MOD;
    }
};

File: 2327. Number of People Aware of a Secret/Solution.py
class Solution:
    def peopleAwareOfSecret_MK1(self, n: int, delay: int, forget: int) -> int:
        MOD = 10**9 + 7
        dp = [0] * n
        dp[0] = 1
        slide_sum = 0
        for i in range(1, n):
            if i >= delay:
                slide_sum += dp[i - delay]
            if i >= forget:
                slide_sum -= dp[i - forget]
            dp[i] = slide_sum % MOD
        return sum(dp[n - forget :]) % MOD

    def peopleAwareOfSecret_MK2(self, n: int, delay: int, forget: int) -> int:
        MOD = 10**9 + 7
        pre_sum = [0] * (n + 1)
        pre_sum[1] = 1
        for i in range(2, n + 1):
            pre_sum[i] = pre_sum[i - 1]
            if i > delay:
                pre_sum[i] += pre_sum[i - delay]
            if i > forget:
                pre_sum[i] -= pre_sum[i - forget]
            pre_sum[i] %= MOD
        return (pre_sum[n] - pre_sum[n - forget]) % MOD
